schema/function.go

```diff
@@ -0,0 +1,326 @@
+		return nil, fmt.Errorf("expected 'any' return type for handler, but got %s", parsedHandler.Type().Out(0))
+	}
+	return &CallableFunctionSchema{
+		IDValue:            id,
+		InputsValue:        inputs,
+		DefaultOutputValue: nil,
+		DisplayValue:       display,
+		Handler:            parsedHandler,
+		DynamicTypeHandler: typeHandler,
+	}, nil
+}
+
+func validateInputTypeCompatibility(
+	inputs []Type,
+	handler reflect.Value,
+) error {
+	// Validate the input types match the provided ones.
+	specifiedParams := len(inputs)
+	actualParams := handler.Type().NumIn()
+	if specifiedParams != actualParams {
+		return fmt.Errorf(
+			"parameter inputs do not match handler inputs. handler has %d, expected %d",
+			actualParams, specifiedParams)
+	}
+	for i := 0; i < len(inputs); i++ {
+		expectedType := inputs[i].ReflectedType()
+		handlerType := handler.Type().In(i)
+		if expectedType != handlerType {
+			return fmt.Errorf(
+				"type mismatch for parameter at index %d. handler has %v, inputs specifies %v",
+				i, handlerType, expectedType)
+		}
+	}
+	return nil
+}
+
+type FunctionSchema struct {
+	IDValue      string  `json:"id"`
+	InputsValue  []Type  `json:"inputs"`
+	OutputValue  Type    `json:"output"`
+	DisplayValue Display `json:"display"`
+}
+
+func (f FunctionSchema) ID() string {
+	return f.IDValue
+}
+
+func (f FunctionSchema) Parameters() []Type {
+	return f.InputsValue
+}
+
+func (f FunctionSchema) Output(_ []Type) (Type, error) {
+	return f.OutputValue, nil
+}
+
+func (f FunctionSchema) Display() Display {
+	return f.DisplayValue
+}
+
+func (f FunctionSchema) String() string {
+	result := f.ID() + "("
+	for i := 0; i < len(f.Parameters()); i++ {
+		if i != 0 {
+			result += ", "
+		}
+		result += string(f.Parameters()[i].TypeID())
+	}
+	result += ") "
+	if f.OutputValue != nil {
+		result += string(f.OutputValue.TypeID())
+	} else {
+		result += "void"
+	}
+	return result
+}
+
+type CallableFunctionSchema struct {
+	IDValue            string  `json:"id"`
+	InputsValue        []Type  `json:"inputs"`
+	DefaultOutputValue Type    `json:"output"`
+	DisplayValue       Display `json:"display"`
+	// Should be a function call with any amount of parameters, and 0, 1 (err or data),
+	// or two return types (err and data).
+	// Params should match types in InputsValue.
+	// Return types should match OutputValue, or not be there if nil. Plus may have one addition return type: error.
+	Handler reflect.Value
+	// Returns the output type based on the input type. For advanced use cases. Cannot be void.
+	DynamicTypeHandler func(inputType []Type) (Type, error)
+}
+
+func (s CallableFunctionSchema) ID() string {
+	return s.IDValue
+}
+
+func (s CallableFunctionSchema) Parameters() []Type {
+	return s.InputsValue
+}
+
+func (s CallableFunctionSchema) Output(inputType []Type) (Type, error) {
+	if s.DynamicTypeHandler == nil {
+		return s.DefaultOutputValue, nil
+	} else {
+		return s.DynamicTypeHandler(inputType)
+	}
+}
```

Comment:
I agree with renaming it to Static. That's definitely the right way to go.

No, it should not use InputsValue. Because inputType is generally supposed to be more specific than InputsValue. That's what makes it dynamic. For example, if the input type specified in the schema is Any, then you can use the dynamic typing to detect int, str, map, etc.